v2/internal/reconcilers/entra/entra_securitygroup_reconciler.go

```diff
@@ -0,0 +1,499 @@
+	if r.canCreate(group) {
+		return r.create(ctx, group, log)
+	}
+
+	// Nothing to do
+	return ctrl.Result{}, nil
+}
+
+func (r *EntraSecurityGroupReconciler) Delete(
+	ctx context.Context,
+	log logr.Logger,
+	eventRecorder record.EventRecorder,
+	obj genruntime.MetaObject,
+) (ctrl.Result, error) {
+	log.V(Status).Info("Updating Entra security group")
+
+	group, err := r.asSecurityGroup(obj)
+	if err != nil {
+		return ctrl.Result{}, eris.Wrapf(err, "creating or updating security group %s", group.Name)
+	}
+
+	// If don't know the Entra ID of the group (captured in an annotation), there's nothing to do.
+	id, ok := getEntraID(obj)
+	if !ok {
+		return ctrl.Result{}, nil
+	}
+
+	client, err := r.EntraClientFactory(ctx, obj)
+	if err != nil {
+		return ctrl.Result{}, eris.Wrap(err, "creating entra client")
+	}
+
+	err = client.Client().Groups().ByGroupId(id).Delete(ctx, nil)
+	if err != nil {
+		// If the group doesn't exist, return nil and nil as we've successfully ensured that it doesn't exist
+		if r.isNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		// If the error is not a 404, return the error
+		var displayName string
+		if group.Spec.DisplayName != nil {
+			displayName = *group.Spec.DisplayName
+		}
+
+		return ctrl.Result{}, eris.Wrapf(err, "failed to delete group %q (%s)", displayName, id)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *EntraSecurityGroupReconciler) Claim(
+	ctx context.Context,
+	log logr.Logger,
+	eventRecorder record.EventRecorder,
+	obj genruntime.MetaObject,
+) error {
+	// Nothing to do
+	//!! Confirm this is true
+	return nil
+}
+
+// update completes our reconciliation by updating an existing Entra security group.
+// ctx is the context for the operation.
+// id is the Entra ID of the group to update.
+// group is the security group to update.
+// eventRecorder is used to record events for the group.
+// log is the logger to use for logging.
+func (r *EntraSecurityGroupReconciler) update(
+	ctx context.Context,
+	id string,
+	group *asoentra.SecurityGroup,
+	log logr.Logger,
+) (ctrl.Result, error) {
+	log = log.WithValues("id", id)
+	log.V(Status).Info("Updating Entra security group")
+
+	// Create our Entra Client
+	client, err := r.EntraClientFactory(ctx, group)
+	if err != nil {
+		return ctrl.Result{}, eris.Wrap(err, "creating entra client prior to update")
+	}
+
+	// Load the existing group by ID
+	g, err := r.loadGroupByID(ctx, id, client.Client())
+	if err != nil {
+		if r.isNotFound(err) {
+			// Group used to exist, but no longer does - it's probably been deleted
+			// Remove the existing annotation and requeue the reconciliation to create a replacement
+			log.V(Status).Info("Group no longer exists")
+			setEntraID(group, "")
+			return ctrl.Result{
+				Requeue: true,
+			}, nil
+		}
+
+		return ctrl.Result{}, eris.Wrapf(err, "getting group by ID %s", id)
+	}
+
+	// Update - PATCH
+	group.Spec.AssignToGroup(g)
+
+	_, err = client.Client().Groups().ByGroupId(id).Patch(ctx, g, nil)
```

Comment:
Do you know what the semantics of this PATCH are? What happens if there's an optional field that the user had previously specified, and then removes from the specification. Now the PATCH will (presumably) not include that field in its payload - will it get cleared or will it be left?

We have some resources in Azure that behave like that too (cough: AKS), but it's not ideal for the IaC style work users will be doing here.

Reply:
If omitted from the payload, the value isn't changed. 

I tried modifying `SecurityGroupSpec.AssignToGroup` to explicitly set every field, but that fails validation - it seems many properties can't be cleared, once set. 

Not sure how we should handle those. Terraform seems to delete/re-create, which seems drastic.

Reply:
Yeah delete+recreate is very awkward for sure, not something we should be doing IMO. Leaving this comment unresolved as it may be worth leaving a comment in the code w.r.t, but it wouldn't be the only resource that has this problem in theory (as I mentioned, AKS does in certain cases too, though usually there there is a way to force the thing off, rather than unsetting it)